packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -460,15 +468,34 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
         document.removeEventListener('keydown', onKeyDown, true);
       }
 
-      if (restoreFocus && nodeToRestore && isElementInScope(document.activeElement, scopeRef.current)) {
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      if (nodeToRestore && restoreFocus && isElementInScope(document.activeElement, scopeRef.current)) {
         requestAnimationFrame(() => {
+          let index = nodeToRestoreArray.lastIndexOf(nodeToRestore);
+          // If the nodeToRestore is the last item in the nodeToRestoreArray.
+          if (index === nodeToRestoreArray.length - 1) {
+            // Iterate backwards through the nodeToRestoreArray until you find a valid node to restore.
+            while (nodeToRestoreArray.length > 0) {
+              let node = nodeToRestoreArray.pop();
+              if (document.body.contains(node)) {
+                focusElement(node);
+                break;
+              }
+            }
+            return;
+          }
+          // Otherwise try to focus the nodeToRestore, and clean up
```

Comment:
i don't quite understand this case, can you explain when we'd hit this?

Reply:
We have to account for FocusScopes where `contain={false}`, as @LFDanLu pointed out in this comment: https://github.com/adobe/react-spectrum/pull/2445#pullrequestreview-977644837

> I'll need to dig into it a some more but the following [story](https://reactspectrum.blob.core.windows.net/reactspectrum/046df7ffb08aff65b9c9db93b89c539be10f4efb/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain&providerSwitcher-toastPosition=bottom) breaks in the following case:
>
>   1. Open the dialog with you keyboard
>   2. Open several layers of nested dialogs via the "open dialog" buttons in each new dialog
>   3. Alt + Tab back to the first "dialog" you opened and hit the "close" button with the enter key
>   4. Note that focus is lost instead of being restored to the first "open dialog" button
>
> The above doesn't reproduce in an earlier version of this PR: https://reactspectrum.blob.core.windows.net/reactspectrum/c465126e8286623894aca7b692cb096121643211/storybook/index.html?path=/story/focusscope--keyboard-navigation-no-contain
> 
> EDIT:
> Seems to be due to something in [a2fad80](https://github.com/adobe/react-spectrum/pull/2445/commits/a2fad80eaffb0724db6b247daefb0e118aea9a51)

Reply:
We iterate backward through the `nodeToRestoreArray` using `pop()` only when the `nodeToRestore` for the closing scope is the last item in the array.